### PR TITLE
Alek/remove to string calls from array instances

### PR DESCRIPTION
**Pull request by AlekSimpson**
## What's changed?
toString should not be called on array instances. RSPEC-2116

## What's your motivation?
[#44](https://github.com/openrewrite/rewrite-static-analysis/issues/44)

## Anything in particular you'd like reviewers to focus on?
I think the recipe is pretty much finished, only two things is there is an edge case with String.format() calls I still haven't quite worked out. It gives an error when I try and run the tests on them. Second thing is that I'm not sure how to get the preconditions to check for if the code is using the array type before it runs. 

## Anyone you would like to review specifically?
@timtebeek 

### Checklist
- [x] I've added unit tests to cover both positive and negative cases
- [x] I've added the license header to any new files through `./gradlew licenseFormat`
- [x] I've used the IntelliJ auto-formatter on affected files
- [ ] I've updated the documentation (if applicable)

### CODE REVIEW

**Comment by timtebeek** [comment]
There's some conflicts here; could you resolve those? Did you perhaps start your branch from your previous branch?

**Comment by AlekSimpson** [comment]
I figured out the edge case and also the precondition check. I think this recipe is functionally complete, but maybe there could be last minute polishes I don't see.

**Review by sambsnyd — APPROVED** [review]
Good work! This will be a useful recipe. 👍 
After you address my feedback this looks good to check in, no need to seek another round of review from me.

File: src/main/java/org/openrewrite/staticanalysis/RemoveToStringCallsFromArrayInstances.java [hunk 1]
```diff
@@ -0,0 +1,116 @@
...
+import org.openrewrite.java.JavaTemplate;
+import org.openrewrite.java.MethodMatcher;
+import org.openrewrite.java.search.UsesMethod;
+import org.openrewrite.java.tree.Expression;
+import org.openrewrite.java.tree.J;
+import org.openrewrite.java.tree.JavaType;
+
+import java.util.List;
+
+public class RemoveToStringCallsFromArrayInstances extends Recipe {
```
[the remark this answers is not in the document]
Comment (knutwannheden) [reply src/main/java/org/openrewrite/staticanalysis/RemoveToStringCallsFromArrayInstances.java:32]: Could be https://rules.sonarsource.com/java/RSPEC-2116

I would also like to propose that we change the name, since we are either replacing `Object#toString()` with `Arrays.toString()` and sometimes we just add the latter. So maybe something like `UseArraysToStringForArrays`?

File: src/main/java/org/openrewrite/staticanalysis/RemoveToStringCallsFromArrayInstances.java [hunk 2]
```diff
@@ -0,0 +1,116 @@
...
+    private static final MethodMatcher TO_STRING_MATCHER = new MethodMatcher("java.lang.Object toString(..)");
```
Comment (knutwannheden) [inline src/main/java/org/openrewrite/staticanalysis/RemoveToStringCallsFromArrayInstances.java:33]:
```suggestion
    private static final MethodMatcher TO_STRING_MATCHER = new MethodMatcher("java.lang.Object toString()");
```

File: src/main/java/org/openrewrite/staticanalysis/RemoveToStringCallsFromArrayInstances.java [hunk 3]
```diff
@@ -0,0 +1,116 @@
...
+    private static final MethodMatcher PRINTLN_MATCHER = new MethodMatcher("java.io.PrintStream println(..)");
```
Comment (knutwannheden) [inline src/main/java/org/openrewrite/staticanalysis/RemoveToStringCallsFromArrayInstances.java:34]: As we only want to match `print(Object)` and `println(Object)` here:
```suggestion
    private static final MethodMatcher PRINTLN_MATCHER = new MethodMatcher("java.io.PrintStream print*(java.lang.Object)");
```

Further, I think there will be quite a few more methods like this one. Some which come to mind: `Objects#toString(Object)`, `String#valueOf(Object)`, `StringBuilder#insert(int, Object)`, `StringBuilder#append(Object)`, etc.

File: src/main/java/org/openrewrite/staticanalysis/RemoveToStringCallsFromArrayInstances.java [hunk 4]
```diff
@@ -0,0 +1,116 @@
...
+    private static final MethodMatcher STR_FORMAT_MATCHER = new MethodMatcher("java.lang.String format(..)");
```
Comment (knutwannheden) [inline src/main/java/org/openrewrite/staticanalysis/RemoveToStringCallsFromArrayInstances.java:35]: We should also cover `java.io.PrintStream#format(String, Object...)`.

File: src/main/java/org/openrewrite/staticanalysis/RemoveToStringCallsFromArrayInstances.java [hunk 5]
```diff
@@ -0,0 +1,116 @@
...
+                Expression select = m.getArguments().get(0);
+                String builder_string = "System.out.println(Arrays.toString(#{anyArray(java.lang.String)}))";
+
+                return buildReplacement(builder_string, m, select);
+            }else if (STR_FORMAT_MATCHER.matches(m)) {
+                List<Expression> arguments = m.getArguments();
+                StringBuilder builder_string = new StringBuilder("String.format(#{any(java.lang.String)}");
+                boolean arrayExists = false;
+
+                for (int i = 0; i<arguments.size(); i++) {
```
Comment (knutwannheden) [inline src/main/java/org/openrewrite/staticanalysis/RemoveToStringCallsFromArrayInstances.java:76]: For reasons of efficiency (and simplicity in a way) I am wondering if we shouldn't instead use cursor messaging and transform this bottom-up. What I mean is that you in this method would add a message to the cursor to indicate that you have found one of the methods which need transforming. After that you would do the `super` call and then you would add an override for `visitExpression()` which in case the parent cursor (skipping over any intermediate `J.Parentheses` cursors) has the mentioned message set and the expression is array-typed would return the array wrapped in `Arrays#toString()`. Does that make sense to you? This logic would then be exactly the same for both `STR_FORMAT_MATCHER` and `PRINTLN_MATCHER` and all the other methods I mentioned.

File: src/main/java/org/openrewrite/staticanalysis/RemoveToStringCallsFromArrayInstances.java [hunk 6]
```diff
@@ -0,0 +1,116 @@
...
+
+    public TreeVisitor<?, ExecutionContext> getVisitor() {
+        return Preconditions.check(Preconditions.or(
+                new UsesMethod<>(TO_STRING_MATCHER),
+                new UsesMethod<>(PRINTLN_MATCHER),
+                new UsesMethod<>(STR_FORMAT_MATCHER)
+        ), new RemoveToStringFromArraysVisitor());
+    }
+
+    private static class RemoveToStringFromArraysVisitor extends JavaIsoVisitor<ExecutionContext> {
```
Comment (knutwannheden) [inline src/main/java/org/openrewrite/staticanalysis/RemoveToStringCallsFromArrayInstances.java:55]: There is one more important override missing here: `visitBinary()`. This is for the case when `+` is being used to concatenate a string with an array.

File: src/main/java/org/openrewrite/staticanalysis/RemoveToStringCallsFromArrayInstances.java (hunk 4 above)
Comment (timtebeek) [reply src/main/java/org/openrewrite/staticanalysis/RemoveToStringCallsFromArrayInstances.java:35]: String.formatted is another option, as well as the printf variants.

File: src/main/java/org/openrewrite/staticanalysis/RemoveToStringCallsFromArrayInstances.java (hunk 1 above)
[the remark this answers is not in the document]
Comment (AlekSimpson) [reply src/main/java/org/openrewrite/staticanalysis/RemoveToStringCallsFromArrayInstances.java:32]: It is https://rules.sonarsource.com/java/RSPEC-2116. My bad for not tagging it initially. That name change makes sense to me.

File: src/main/java/org/openrewrite/staticanalysis/RemoveToStringCallsFromArrayInstances.java (hunk 5 above)
Comment (AlekSimpson) [reply src/main/java/org/openrewrite/staticanalysis/RemoveToStringCallsFromArrayInstances.java:76]: Yep this makes sense. I'll get started on this change.

**Comment by AlekSimpson** [comment]
Ok, so I've tried to incorporate all the feedback from this thread. I still need to add more tests for a bunch of the other methods mentioned above. However, I think I should wait to do that till I have the cursor messaging working. I've updated the branch with my implementation of the cursor messaging but it doesn't work. The message never seems to be found in the `visitExpression()` visitor and I'm not sure why.

**Comment by knutwannheden** [comment]
> Ok, so I've tried to incorporate all the feedback from this thread. I still need to add more tests for a bunch of the other methods mentioned above. However, I think I should wait to do that till I have the cursor messaging working. I've updated the branch with my implementation of the cursor messaging but it doesn't work. The message never seems to be found in the `visitExpression()` visitor and I'm not sure why. @knutwannheden

@AlekSimpson You got it almost all right! The important thing here is that the `super.visitMethodInvocation()` call comes after the message is added to the cursor. Otherwise `visitExpression()` will end up getting called before the cursor message is added. I made some changes and will let you take it from here.

**Comment by AlekSimpson** [comment]
All changes have been made and tests have been updated. All tests pass.

File: src/main/java/org/openrewrite/staticanalysis/RemoveToStringCallsFromArrayInstances.java [hunk 7]
```diff
@@ -0,0 +1,132 @@
...
+
+        @Override
+        public J.Binary visitBinary(J.Binary binary, ExecutionContext ctx) {
+            Expression left = binary.getLeft();
+            Expression right = binary.getRight();
+
+            Cursor c = getCursor().dropParentWhile(is -> is instanceof J.Parentheses || !(is instanceof Tree));
+            if (c.getValue() instanceof J.MethodInvocation) {
+                if (PRINT_MATCHER.matches((J.MethodInvocation) c.getValue())) {
+                    if (left.getType() instanceof JavaType.Array || right.getType() instanceof JavaType.Array) {
```
Comment (knutwannheden) [inline src/main/java/org/openrewrite/staticanalysis/RemoveToStringCallsFromArrayInstances.java:123]: There is no need to constrain this to method invocations. So you can drop the two outer `if`s here and on this innermost `if` I would first also check the condition `binary.getOperator() == J.Binary.Type.Addition`.

File: src/main/java/org/openrewrite/staticanalysis/RemoveToStringCallsFromArrayInstances.java [hunk 8]
```diff
@@ -0,0 +1,132 @@
...
+                }
+            }
+
+            return super.visitMethodInvocation(mi, ctx);
+        }
+
+        @Override
+        public Expression visitExpression(Expression exp, ExecutionContext ctx) {
+            Expression e = (Expression) super.visitExpression(exp, ctx);
+            if (e.getType() instanceof JavaType.Array) {
```
Comment (knutwannheden) [inline src/main/java/org/openrewrite/staticanalysis/RemoveToStringCallsFromArrayInstances.java:101]: When you add a test like `System.out.println("" + (arr))` (where `arr` is an array), you will notice that you get a runtime exception. That is because `J.Parentheses` doesn't extend `TypedTree` (and I currently can't tell you if it should or not). To fix this my suggestion is to simply add an `instanceof` check:

```suggestion
            if (e instanceof TypedTree && e.getType() instanceof JavaType.Array) {
```

**Review by knutwannheden — COMMENTED** [review]
Almost there! I added a few more comments which should be addressed.

File: src/main/java/org/openrewrite/staticanalysis/RemoveToStringCallsFromArrayInstances.java [hunk 9]
```diff
@@ -0,0 +1,132 @@
...
+
+import java.util.Arrays;
+import java.util.Collections;
+import java.util.List;
+import java.util.Set;
+import java.util.stream.Collectors;
+
+public class RemoveToStringCallsFromArrayInstances extends Recipe {
+    private static final MethodMatcher OBJECT_TOSTRING_MATCHER = new MethodMatcher("java.lang.Object toString()");
+    private static final MethodMatcher OBJECTS_TOSTRING_MATCHER = new MethodMatcher("java.lang.Objects toString()");
```
[the remark this answers is not in the document]
Comment (AlekSimpson) [reply src/main/java/org/openrewrite/staticanalysis/RemoveToStringCallsFromArrayInstances.java]: The latest changes should address this now.

File: src/main/java/org/openrewrite/staticanalysis/RemoveToStringCallsFromArrayInstances.java [hunk 10]
```diff
@@ -98,7 +98,7 @@ public J visitMethodInvocation(J.MethodInvocation mi, ExecutionContext ctx) {
         @Override
         public Expression visitExpression(Expression exp, ExecutionContext ctx) {
             Expression e = (Expression) super.visitExpression(exp, ctx);
-            if (e.getType() instanceof JavaType.Array) {
+            if (e instanceof TypedTree && e.getType() instanceof JavaType.Array) {
```
[the remark this answers is not in the document]
Comment (knutwannheden) [reply src/main/java/org/openrewrite/staticanalysis/RemoveToStringCallsFromArrayInstances.java:111]: We could replace `e.getType() instanceof JavaType.Array` with `TypeUtils.asArray(e.getType()) != null` but I don't think it is really much better. IMHO the `instanceof` here is fine.

File: src/main/java/org/openrewrite/staticanalysis/RemoveToStringCallsFromArrayInstances.java [hunk 11]
```diff
@@ -117,13 +127,8 @@ public J.Binary visitBinary(J.Binary binary, ExecutionContext ctx) {
...
             Expression right = binary.getRight();
 
-            Cursor c = getCursor().dropParentWhile(is -> is instanceof J.Parentheses || !(is instanceof Tree));
-            if (c.getValue() instanceof J.MethodInvocation) {
-                if (PRINT_MATCHER.matches((J.MethodInvocation) c.getValue())) {
-                    if (left.getType() instanceof JavaType.Array || right.getType() instanceof JavaType.Array) {
-                        getCursor().putMessage("BINARY_FOUND", binary);
-                    }
-                }
+            if (binary.getOperator() == J.Binary.Type.Addition || left.getType() instanceof JavaType.Array || right.getType() instanceof JavaType.Array) {
```
Comment (knutwannheden) [inline src/main/java/org/openrewrite/staticanalysis/RemoveToStringCallsFromArrayInstances.java:130]: The binary must be using `+`:
```suggestion
            if (binary.getOperator() == J.Binary.Type.Addition && (left.getType() instanceof JavaType.Array || right.getType() instanceof JavaType.Array)) {
```

**Review by knutwannheden — COMMENTED** [review]
Apart from one last minor comment I think this looks good.

**Review by timtebeek — APPROVED** [review]
